src/consensus/params.cpp

```diff
@@ -36,32 +39,147 @@ namespace Consensus {
+     *
+     * Previous implementations of this logic were specialized to the
+     * first halving.
+     */ 
+    int Params::HalvingHeight(int nHeight, int halvingIndex) const {
         // zip208
-        // FoundersRewardLastBlockHeight := max({ height ⦂ N | Halving(height) < 1 })
-        // Halving(h) is defined as floor(f(h)) where f is a strictly increasing rational
-        // function, so it's sufficient to solve for f(height) = 1 in the rationals and
-        // then take ceiling(height - 1).
-        // H := blossom activation height; SS := SubsidySlowStartShift(); R := BLOSSOM_POW_TARGET_SPACING_RATIO
+        // HalvingHeight(i) := max({ height ⦂ N | Halving(height) < i }) + 1
+        //
+        // Halving(h) returns the halving index at the specified height.  It is
+        // defined as floor(f(h)) where f is a strictly increasing rational
+        // function, so it's sufficient to solve for f(height) = halvingIndex
+        // in the rationals and then take ceiling(height - 1).
+        //
+        // H := blossom activation height; 
+        // SS := SubsidySlowStartShift(); 
+        // R := 1 / (postInterval / preInterval) = BLOSSOM_POW_TARGET_SPACING_RATIO
+        // (The following calculation depends on BLOSSOM_POW_TARGET_SPACING_RATIO being an integer.) 
+        //
         // preBlossom:
         // 1 = (height - SS) / preInterval
         // height = preInterval + SS
+        //
         // postBlossom:
-        // 1 = (H - SS) / preInterval + (height - H) / postInterval
-        // height = H + postInterval - (H - SS) * (postInterval / preInterval)
-        // height = H + postInterval - (H - SS) * R
-        // Note: This depends on R being an integer
-        bool blossomActive = NetworkUpgradeActive(nHeight, Consensus::UPGRADE_BLOSSOM);
-        if (blossomActive) {
+        // i = (H - SS) / preInterval + (HalvingHeight(i) - H) / postInterval
+        // preInterval = postInterval / R
+        // i = (H - SS) / (postInterval / R) + (HalvingHeight(i) - H) / postInterval
+        // i = (R * (H - SS) + HalvingHeight(i) - H) / postInterval
+        // HalvingHeight(i) = postInterval * i - R * (H - SS) + H  
+        // under normal circumstances, Halving(H) = 0
+        if (NetworkUpgradeActive(nHeight, Consensus::UPGRADE_BLOSSOM)) {
             int blossomActivationHeight = vUpgrades[Consensus::UPGRADE_BLOSSOM].nActivationHeight;
-            // The following calculation depends on BLOSSOM_POW_TARGET_SPACING_RATIO being an integer. 
-            return blossomActivationHeight + nPostBlossomSubsidyHalvingInterval
-                - (blossomActivationHeight - SubsidySlowStartShift()) * BLOSSOM_POW_TARGET_SPACING_RATIO - 1;
+            return blossomActivationHeight 
+                - ((blossomActivationHeight - SubsidySlowStartShift()) * BLOSSOM_POW_TARGET_SPACING_RATIO)
+                + (nPostBlossomSubsidyHalvingInterval * (halvingIndex - Halving(blossomActivationHeight)));
```

Comment:
Reorganise this to follow the comment:
```suggestion
            return (nPostBlossomSubsidyHalvingInterval * (halvingIndex - Halving(blossomActivationHeight)))
                - (BLOSSOM_POW_TARGET_SPACING_RATIO * (blossomActivationHeight - SubsidySlowStartShift()))
                + blossomActivationHeight;
```

Reply:
@daira and I also think that `- Halving(blossomActivationHeight)` shouldn't be necessary.